File: src/main/ebay/media.ts
/** Image upload to eBay Picture Services (Media API). */

import { readFile } from 'node:fs/promises'
import { basename } from 'node:path'

import { mediaBase, type AppConfig } from '@main/config'

import { EbayApiError, type TokenSource } from './client'

const UPLOAD_TIMEOUT_MS = 60_000

/** Upload a local image; returns the eBay-hosted image URL. */
export async function uploadImage(
  config: AppConfig,
  auth: TokenSource,
  path: string,
): Promise<string> {
  const token = await auth.getUserToken()
  const bytes = await readFile(path)

  const form = new FormData()
  form.append('image', new Blob([bytes]), basename(path))

  const response = await fetch(`${mediaBase(config)}/commerce/media/v1/image`, {
    method: 'POST',
    headers: { Authorization: `Bearer ${token}` },
    body: form,
    signal: AbortSignal.timeout(UPLOAD_TIMEOUT_MS),
  })

  if (!response.ok) {
    throw new EbayApiError(
      response.status,
      `Bild-Upload fehlgeschlagen: ${(await response.text()).slice(0, 300)}`,
    )
  }

  // The created image URL comes back either in the body or via Location.
  const text = await response.text()
  if (text) {
    const body = JSON.parse(text) as { imageUrl?: string }
    if (body.imageUrl) return body.imageUrl
  }

  const location = response.headers.get('Location')
  if (!location) {
    throw new EbayApiError(response.status, 'eBay hat keine Bild-URL zurückgegeben.')
  }

  const imageId = location.replace(/\/$/, '').split('/').pop()
  const info = await fetch(`${mediaBase(config)}/commerce/media/v1/image/${imageId}`, {
    headers: { Authorization: `Bearer ${token}` },
    signal: AbortSignal.timeout(UPLOAD_TIMEOUT_MS),
  })
  const parsed = (await info.json()) as { imageUrl?: string }
  if (!parsed.imageUrl) {
    throw new EbayApiError(info.status, 'eBay hat keine Bild-URL zurückgegeben.')
  }
  return parsed.imageUrl
}

File: src/epay_tool/ebay/media.py
"""Image upload to eBay Picture Services (Media API)."""

from __future__ import annotations

from pathlib import Path

import httpx

from epay_tool.ebay.client import EbayApiError, EbayClient

# The Media API lives on a separate host than the other REST APIs.
_MEDIA_HOSTS = {
    True: "https://apim.ebay.com",           # production
    False: "https://apim.sandbox.ebay.com",  # sandbox
}


def upload_image(client: EbayClient, path: Path) -> str:
    """Upload a local image; returns the eBay-hosted image URL."""
    host = _MEDIA_HOSTS[client.settings.is_production]
    token = client.auth.user_token()
    with path.open("rb") as fh:
        resp = httpx.post(
            f"{host}/commerce/media/v1/image",
            files={"image": (path.name, fh, "image/jpeg")},
            headers={"Authorization": f"Bearer {token}"},
            timeout=60,
        )
    if resp.status_code not in (200, 201):
        raise EbayApiError(resp.status_code, f"Image upload failed: {resp.text[:300]}")
    # The created image URL is returned in the Location header / body.
    if "Location" in resp.headers:
        image_id = resp.headers["Location"].rstrip("/").rsplit("/", 1)[-1]
        info = client.get(f"/commerce/media/v1/image/{image_id}")
        return info.get("imageUrl", "")
    return resp.json().get("imageUrl", "")
